### PR TITLE
Issue #460 Add GitLab CI runner

### PR DESCRIPTION
**Pull request by gbraad**
Added GitLab CI runner definition to build on a GitLab instance. Example:
  * https://gitlab.com/gbraad/minishift/-/jobs/28627751
 * https://gitlab.com/minishift/minishift/-/jobs/28634280

`oc_env_test.go` needed to be given a forced shell, else the test fails with:
```
=== RUN   Test_unix_oc_path
The default lines below are for a sh/bash shell, you can specify the shell you're using, with the --shell flag.

--- FAIL: Test_unix_oc_path (0.00s)
	oc_env_test.go:28: Unexepcted error: Error: Unknown shell
```
Note: running with `bash -c "make test"` did not make a difference

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    - ln -s $CURRENT_BUILD_PATH $HOME/gopath/src/github.com/minishift/minishift
+    - cd $HOME/gopath/src/github.com/minishift/minishift/
+    - make prerelease
+    - rm -rf $HOME/gopath/src/github.com/minishift/minishift/out/bindata
+    - mv $HOME/gopath/src/github.com/minishift/minishift/out/ $CURRENT_BUILD_PATH/public
+  artifacts:
+    paths:
+      - public
+
+# git is needed for glide
```
Comment (budhrg) [inline .gitlab-ci.yml]: This looks like random comment and you forgot to remove it.

File: cmd/minishift/cmd/oc_env_test.go [hunk 2]
```diff
@@ -23,7 +23,7 @@ import (
 )
 
 func Test_unix_oc_path(t *testing.T) {
-	shellConfig, err := getOcShellConfig("/Users/john/.minishift/cache/oc/v1.5.0/oc", "")
+	shellConfig, err := getOcShellConfig("/Users/john/.minishift/cache/oc/v1.5.0/oc", "bash")
```
Comment (budhrg) [inline cmd/minishift/cmd/oc_env_test.go]: Fair enough.

File: .gitlab-ci.yml (hunk 1 above)
Comment (gbraad) [inline .gitlab-ci.yml]: It is not random comment. It seems git is needed for glide to work. but it can be removed.

**Comment by praveenkumar** [comment]
@gbraad Can you explain what value we are going to add with this PR? Are we going to add another CI on gitlab side?

**Comment by gbraad** [comment]
I use GitLab myself a lot. For me, this means I do not have to use a build-wrapper which I manually trigger.

For community. this would mean they can easily set up a complete environment that can perform the same builds as we do, without using 3rd party services.

**Comment by gbraad** [comment]
@praveenkumar none of the builds from yesterday mentioned this: https://travis-ci.org/minishift/minishift/builds/261835519

**Comment by praveenkumar** [comment]
> None of the builds from yesterday mentioned

https://github.com/minishift/minishift/commit/9d36b5c261daec20dcdda3539b9d391e56bf225a this is where it happen. Yesterday @budhrg and @hferentschik discussing it not sure if that merged yestereday or on Friday.

**Comment by gbraad** [comment]
I mean, they do not report it in their errors/output when building. While `make` on Fedora reported:
```
make: Nothing to be done for 'fmtcheck'.
make: Nothing to be done for 'cross'.
make: Nothing to be done for 'test'.
```

**Comment by hferentschik** [comment]
> For community. this would mean they can easily set up a complete environment that can perform the same builds as we do, without using 3rd party services.

I am confused. What do you mean? Why are you saying no 3rd party services? They would be using hosted GitLab, right? Or are you expecting them to install GitLab locally? 

Regarding same environment and build, if you fork the minishift repo you automatically get for example Travis builds. What is the benefit for your of GitLab over any of the configured CI systems. Each of then allows you to build Minishift from your own fork? (just trying to understand)

**Comment by gbraad** [comment]
Hosted or on-premise makes no difference in how you can use GitLab. For
hosted you can use shared or personal runners.

**Comment by gbraad** [comment]
What is the harm of adding this definition file?

Note: maintenance is not an issue as I will run it daily.

**Comment by hferentschik** [comment]
> What is the harm of adding this definition file?

No harm directly, just adding more "code" which needs to be "understood". I am neutral on this one. I just wanted to understand where you are coming from and why the existing options are not "working" for you.

**Comment by gbraad** [comment]
the other options are working, but not a preference. I use a GitLab instance at home and the hosted.

At the moment I have to resolve to a build-wrapper. This would not be a problem if GitLab would have resolved: https://gitlab.com/gitlab-org/gitlab-ce/issues/15041, but until now they do not allow external build definitions.

Builds here runs with the provided RPMs from Fedora's repository for `golang` and `glide`.

**Comment by gbraad** [comment]
Note: the understanding of the code could be simplified if Travis and GitLab share the same build script,

**Comment by hferentschik** [comment]
@minishift/minishift-dev WDYT?

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+build:
+  image: fedora:26
```
Comment (LalatenduMohanty) [inline .gitlab-ci.yml]: 👍

**Comment by praveenkumar** [comment]
@hferentschik I don't see any issue for merging this PR if that helps any way to community.

**Comment by LalatenduMohanty** [comment]
I was going through https://about.gitlab.com/features/gitlab-ci-cd/ and talked to @gbraad about it.  The USP of gitlab is that it can be used on the gitlab cloud as well with local gitlab instance. For example we can run the CI on a local machine through gitlab. I think that can come handy in future. Also as per PR it runs on fedora and thats a added advantage. May be we can try running it on CentOS too. 

I understand and share the concern that we will have 5 CI instances and that would be a PITA to maintain in long run.  The question is should we have travis or gitlab? At this point I am not sure how stable is gitlab but Travis has this limitation that we can only run on top of Ubuntu. So lets try both and after 1 month lets discuss if we want to remove one of them.

**Comment by praveenkumar** [comment]
> I understand and share the concern that we will have 5 CI instances and that would be a PITA to maintain in long run. The question is should we have travis or gitlab? At this point I am not sure how stable is gitlab but Travis has this limitation that we can only run on top of Ubuntu. So lets try both and after 1 month lets discuss if we want to remove one of them.

I don't think we can have gitlab-ci running for a github repo https://forum.gitlab.com/t/is-there-a-way-to-trigger-a-gitlab-ci-build-from-github/2276/3 so having it 5 CI is out of question right now. What I understand is if a user using own private gitlab instance then for him it is easy to use gitlab ci which this PR make use of. 

@gbraad Did I understand it correct?

**Comment by gbraad** [comment]
It is possible to trigger a build using a webhook from GitHub. The build definition for this would also have to change a little and would copy the build-wrapper idea I used before. I would not recommend going down this road. At the moment I do not say it is one of the main CI pipelines yet, and for now not going to replace Travis, but maybe it is a noble goal to attempt this over time. And I am more than happy to take this as a side project

**Comment by hferentschik** [comment]
> I don't think we can have gitlab-ci running for a github repo https://forum.gitlab.com/t/is-there-a-way-to-trigger-a-gitlab-ci-build-from-github/2276/3 so having it 5 CI is out of question right now. 

+1, that's also my understanding. This is just about adding the config that it can be used, not about enabling another CI build integrated into GitHub.

> What I understand is if a user using own private gitlab instance then for him it is easy to use gitlab ci which this PR make use of.

+1

**Comment by hferentschik** [comment]
> and for now not going to replace Travis

+1 Travis is a de facto standard of course. I would not like to replace it.

**Pull request merged**